src/golang/lib/workflow/artifact/migrate.go

```diff
@@ -0,0 +1,101 @@
+package artifact
+
+import (
+	"context"
+
+	"github.com/aqueducthq/aqueduct/lib/collections/artifact"
+	"github.com/aqueducthq/aqueduct/lib/collections/artifact_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/shared"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow_dag"
+	"github.com/aqueducthq/aqueduct/lib/database"
+	"github.com/aqueducthq/aqueduct/lib/storage"
+	"github.com/aqueducthq/aqueduct/lib/workflow/utils"
+	"github.com/sirupsen/logrus"
+)
+
+// Migrate moves from workflow artifact result content from `oldConf` to `newConf`.
+func Migrate(
+	ctx context.Context,
+	oldConf *shared.StorageConfig,
+	newConf *shared.StorageConfig,
+	dagReader workflow_dag.Reader,
+	dagWriter workflow_dag.Writer,
+	artifactReader artifact.Reader,
+	artifactResultReader artifact_result.Reader,
+	db database.Database,
+) error {
+	// Wait until there are no more workflow runs in progress
+	lock := utils.NewExecutionLock()
+	if err := lock.Lock(); err != nil {
+		return err
+	}
+	defer func() {
+		unlockErr := lock.Unlock()
+		if unlockErr != nil {
+			logrus.Errorf("Unexpected error when unlocking workflow execution lock: %v", unlockErr)
+		}
+	}()
+
+	oldStore := storage.NewStorage(oldConf)
+	newStore := storage.NewStorage(newConf)
+
+	txn, err := db.BeginTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer database.TxnRollbackIgnoreErr(ctx, txn)
+
+	dags, err := dagReader.GetWorkflowDags(ctx, nil, txn)
+	if err != nil {
+		return err
+	}
+
+	for _, dag := range dags {
```

Comment:
If the a migration fails in the middle, and we try to run it again, we'll be able to detect already-migrated dags by checking it's storage config column right? We can then skip migrating those dags.

Reply:
The migration is atomic, so even if the content has been copied over, the DB records will not have been updated, so we simply run the entire migration again.